askpass: Fix cli path when executed in a remote server

marcocondrache opened the pull request:
Closes #39469
Closes #39438
Closes #39458

I'm not able to test it, i would appreciate if somebody could do it. I think this bug was present also for SSH remote projects

Release Notes:

- Fixed an issue where zed bin was not found in remote servers for askpass

marcocondrache committed:
fix: when askpass is executed in remote server we should use the executable itself

Signed-off-by: Marco Mihai Condrache <52580954+marcocondrache@users.noreply.github.com>

marcocondrache commented:
I just noticed that `get_shell_safe_zed_path` it's marked unix only

marcocondrache committed:
fix: this should be unix safe

Signed-off-by: Marco Mihai Condrache <52580954+marcocondrache@users.noreply.github.com>

marcocondrache committed:
chore: this should make it platform agnostic

Signed-off-by: Marco Mihai Condrache <52580954+marcocondrache@users.noreply.github.com>

maxbrunsfeld reviewed (dismissed):
@marcocondrache Thanks so much for tracking this down. Nice work.

I would love to solve this in a way that doesn't rely on checking the filename of the current exe.

Here's one way to fix it that I think is cleaner. Let me know if this makes sense to you:

* We should put back the `--askpass` argument in the main editor on linux and macOS.
* There should be a static variable in `ask_pass.rs` called `ASKPASS_PROGRAM` (a `PathBuf`) - the path to the program that should be invoked for askpass. By default, `ASKPASS_PROGRAM` should be initialized to the `current_exe()`. On mac and Linux, and in the remote server, it should just be left that way.
* The `askpass` module should have a public setter function called `set_askpass_program` that can override the `ASKPASS_PROGRAM`. We should only call that function on Windows, in `zed::main` (passing in the path to the CLI).

marcocondrache committed:
fix: restore askpass argument

Signed-off-by: Marco Mihai Condrache <52580954+marcocondrache@users.noreply.github.com>

marcocondrache committed:
fix: use static variable and extract shell safe logic

Signed-off-by: Marco Mihai Condrache <52580954+marcocondrache@users.noreply.github.com>

marcocondrache committed:
fix: retrieve cli path

Signed-off-by: Marco Mihai Condrache <52580954+marcocondrache@users.noreply.github.com>

marcocondrache commented:
@maxbrunsfeld I’ve updated the PR with the changes. I moved the logic for turning a path into a shell-safe string into the `PathExt` trait, since it felt generic enough to live there but if you think it makes more sense somewhere else, I’m happy to move it.

I’m also not too sure about the `get_zed_cli_path` function since it's called for `set_askpass_program`. Maybe it’d be simpler to inline that directly in `zed::main`, but I’ll let you guide what feels best here.

For testing, I can only try things out on macOS. I’ll give it a spin there, but unfortunately I won’t be much help on Linux or Windows 😞

osiewicz commented on crates/util/src/util.rs:
```diff
@@ -288,27 +290,21 @@ fn load_shell_from_passwd() -> Result<()> {
+        .to_string())
 }
 
 /// Returns a shell escaped path for the zed cli executable, this function
```
Nit: is this doc comment outdated? The returned path is no longer shell-escaped.

osiewicz commented on crates/askpass/src/askpass.rs:
```diff
@@ -249,12 +262,18 @@ pub fn main(socket: &str) {
-        "{shebang}\n{print_args} | {zed_cli} --askpass={askpass_socket} 2> /dev/null \n",
-        zed_cli = zed_cli_path,
+        "{shebang}\n{print_args} | {askpass_program} --askpass={askpass_socket} 2> /dev/null \n",
+        askpass_program = askpass_program,
```
```suggestion
```

osiewicz commented on crates/util/src/paths.rs:
```diff
@@ -153,6 +154,29 @@ impl<T: AsRef<Path>> PathExt for T {
+            // As of writing, this can only be fail if the path contains a null byte, which shouldn't be possible
+            // but shlex has annotated the error as #[non_exhaustive] so we can't make it a compile error if other
+            // errors are introduced in the future :(
+            Ok(shlex::try_quote(path_str)?.to_string())
```
```suggestion
            Ok(shlex::try_quote(path_str)?.into())
```

I'm not sure if that'll compile from the get-go, but `to_string` can be slower than `String::from`. Not that it matters though, so feel free to disregard.

osiewicz replied:
Ah, in that case you can even go the extra mile and use `into_owned`.

marcocondrache committed:
fix: comments

Signed-off-by: Marco Mihai Condrache <52580954+marcocondrache@users.noreply.github.com>

marcocondrache commented on crates/askpass/src/askpass.rs:
```diff
@@ -249,12 +262,18 @@ pub fn main(socket: &str) {
 
+pub fn set_askpass_program(path: std::path::PathBuf) {
+    if ASKPASS_PROGRAM.set(path).is_err() {
+        panic!("askpass program has already been set");
```
Yep that's super right. I missed the debug_panic macro

marcocondrache commented on crates/askpass/src/askpass.rs:
```diff
@@ -249,12 +262,18 @@ pub fn main(socket: &str) {
     }
 }
 
+pub fn set_askpass_program(path: std::path::PathBuf) {
```
I think calling it `init` might be misleading in this case, since it only runs on Windows. Making it run unconditionally could also introduce security concerns - on Unix platforms, `current_exe` should be called right before use to avoid potential privilege escalation.

marcocondrache committed:
fix: missing log crate

Signed-off-by: Marco Mihai Condrache <52580954+marcocondrache@users.noreply.github.com>

marcocondrache committed:
fix: remove trim

Signed-off-by: Marco Mihai Condrache <52580954+marcocondrache@users.noreply.github.com>

marcocondrache committed:
fix: clippy

Signed-off-by: Marco Mihai Condrache <52580954+marcocondrache@users.noreply.github.com>

marcocondrache committed:
fix: machete

Signed-off-by: Marco Mihai Condrache <52580954+marcocondrache@users.noreply.github.com>

osiewicz reviewed (approved):
LGTM! Will give it a go with other platforms before merging.

marcocondrache committed:
fix: comments

Signed-off-by: Marco Mihai Condrache <52580954+marcocondrache@users.noreply.github.com>

marcocondrache commented on crates/util/src/paths.rs:
```diff
@@ -153,6 +154,29 @@ impl<T: AsRef<Path>> PathExt for T {
+
+    /// Try to make a shell-safe representation of the path.
+    ///
+    /// For Unix, the path is escaped to be safe for POSIX shells
```
I've moved the doc comments to the trait definition
